src/App.tsx

```diff
@@ -1,24 +1,161 @@
-  if (!USER_ID) {
-    return <UserWarning />;
-  }
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [sortType, setSortType] = useState<SortType>(SortType.ALL);
+  const [errorMessage, setErrorMessage]
+    = useState<ErrorMessage>(ErrorMessage.NONE);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [loadingTodoId, setLoadingTodoId] = useState(0);
+
+  const activeTodosLength = todos.filter((todo) => !todo.completed).length;
+  const completedTodosLength = todos.length - activeTodosLength;
+
+  const visibleTodos = useMemo(() => {
+    return (todos.filter((todo) => {
+      switch (sortType) {
+        case SortType.ACTIVE:
+          return !todo.completed;
+        case SortType.COMPLETE:
+          return todo.completed;
+        default:
+          return true;
+      }
+    })
+    );
+  }, [sortType, todos]);
+
+  useEffect(() => {
+    const fetchData = async () => {
+      try {
+        const todosFromServer = await getTodos(USER_ID);
+
+        setTodos(todosFromServer);
+      } catch (error) {
+        setErrorMessage(ErrorMessage.LOAD);
+      }
+    };
+
+    fetchData();
+  }, []);
+
+  const handleAddTodo = (title: string) => {
+    if (title.length === 0) {
+      setErrorMessage(ErrorMessage.EMPTY);
+      setTimeout(() => setErrorMessage(ErrorMessage.NONE), 3000);
+    } else {
+      const newTodo = {
+        id: 0,
+        userId: USER_ID,
+        completed: false,
+        title,
+      };
+
+      setTempTodo(newTodo);
+      setErrorMessage(ErrorMessage.NONE);
+
+      addTodo(USER_ID, newTodo)
+        .then((todo) => {
+          setTodos((prevTodos) => {
+            return [...prevTodos, todo];
+          });
+        })
+        .catch(() => {
+          setErrorMessage(ErrorMessage.ADD);
+          setTimeout(() => setErrorMessage(ErrorMessage.NONE), 3000);
+        })
+        .finally(() => {
+          setTempTodo(null);
+        });
+    }
+  };
+
+  const handleDeleteTodo = useCallback(
+    (id: number) => {
+      setLoadingTodoId(id);
+
+      return deleteTodo(id)
+        .then(() => {
+          setTodos((prevTodos) => {
+            return prevTodos.filter((todo) => todo.id !== id);
+          });
+        })
+        .catch(() => {
+          setErrorMessage(ErrorMessage.DELETE);
+          setTimeout(() => setErrorMessage(ErrorMessage.NONE), 3000);
+        })
+        .finally(() => {
+          setTempTodo(null);
+          setLoadingTodoId(0);
+        });
+    },
+    [deleteTodo],
+  );
+
+  const handleClearCompleted = () => {
+    todos.filter((todo) => todo.completed).forEach((todo) => {
+      handleDeleteTodo(todo.id)
+        .then(() => setTodos(todos.filter(({ completed }) => !completed)));
+    });
+  };
+
+  const handleToggleTodo = (id: number) => {
+    const todoCheck = todos.find((todo) => todo.id === id) as Todo;
+
+    todoCheck.completed = !todoCheck?.completed;
+
+    const notChanged = todos.filter(todo => todo.id !== id);
```

Comment:
instead of two iterations, you may use `map` method, and inside callback check id. If the same - return _copy_ of todo with reversed `compleated` property, otherwise return the same, not changed todo

P.S. you may even do it inside setTodos: `setTodos((todos) => todos.map(// callback here))`

Reply:
handleClearCompleted fixed